### ISSUE TITLE
[Blackhole Bringup] Reshuffle Rows Support

### ISSUE THREAD

**Issue by ncvetkovicTT** [tenstorrent/tt-metal#11816]
As a part of #10226 we need to enable several functionalities on BH, one of which is reshuffle rows. On BH submodule side this requires including `sfpu/ckernel_sfpu_reshuffle_rows.h" ` inside `tt_metal/third_party/tt_llk_blackhole/common/inc/ckernel_sfpu.h` and adding the file itself, which contains LLK implementations.  On metal side, we need to add `ckernel_sfpu_reshuffle_rows.h` and `llk_math_eltwise_unary_sfpu_reshuffle_rows.h` API files. 

In order to run the tests for BH, we need to enable the arch in `embedding_backward_device_operation.cpp`, but once we run `tests/ttnn/unit_tests/operations/backward/test_backward_embedding.py` test, several tests are skipped while others fail due to failed atol/rtol/pcc value. The issue can be reproduced using the following branches:
tt-metal: [ncvetkovic/10226-bh_new_sfpu_functions](https://github.com/tenstorrent/tt-metal/tree/ncvetkovic/10226-bh_new_sfpu_functions)
tt-llk-bh: ncvetkovic/10226-bh_new_sfpu_functions (waiting for permissions to be able to push)

Error example:
```
tests.ttnn.unit_tests.operations.backward.test_backward_embedding:test_embedding_bw:67 - Max ATOL Delta: 100.0, Max RTOL Delta: nan, PCC: 0.00976503487902343, PCC check failed
```

```      
def test_embedding_bw(input_dtype, output_dtype, batch_size, seq_len, embedding_dim, num_embeddings, device):
      torch.manual_seed(1234)
      
      if input_dtype == ttnn.bfloat16 and num_embeddings > 256:
          pytest.skip("Skipping tests with large vocab sizes for bfloat16 indices!")
      
      input_shape = (batch_size, seq_len)
      input_index = torch.randint(0, num_embeddings, input_shape)
      input_tensor = ttnn.from_torch(input_index, dtype=input_dtype, device=device)
      
      weights_shape = (num_embeddings, embedding_dim)
      weights = torch.randn(weights_shape, requires_grad=True)
      weights_ttnn = ttnn.from_torch(weights, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
      
      grad_shape = (1, 1, batch_size * seq_len, embedding_dim)
      grad_data = torch.randn(grad_shape, requires_grad=True)
      grad_tensor = ttnn.from_torch(grad_data, dtype=output_dtype, layout=ttnn.TILE_LAYOUT, device=device)
      
      tt_output_tensor_on_device = ttnn.embedding_bw(input_tensor, weights_ttnn, grad_tensor, dtype=output_dtype)
      tt_output_tensor = ttnn.to_torch(tt_output_tensor_on_device)
      
      # PyTorch reference
      weights.retain_grad()
      pyt_y = torch.nn.functional.embedding(input_index, weights).reshape(grad_shape)
      pyt_y.backward(gradient=grad_data)
      golden_output_tensor = weights.grad
      
      comp_pass, comp_out = comparison_funcs.comp_pcc(golden_output_tensor, tt_output_tensor)  
      logger.debug(comp_out)
>    assert comp_pass
E    assert False
tests/ttnn/unit_tests/operations/backward/test_backward_embedding.py:68: AssertionError
```

**Comment by cmaryanTT** [comment]
@vmilicevicTT note the increased priority. This is gating  BH training.

**Comment by vmilicevicTT** [comment]
@cmaryanTT , is this now a P1?  Please re-prioritize it on your board along with other P1s.  Thanks!

**Comment by dmakoviichuk-tt** [comment]
hi @vmilicevicTT  @cmaryanTT  do you have any ETA?
Thanks

**Comment by ncvetkovicTT** [comment]
@dmakoviichuk-tt I'll give an ETA as soon as I take a look tomorrow.

**Comment by ncvetkovicTT** [comment]
ETA: ~12. Sept. given that the LLK team has limited availability next week.

**Comment by ncvetkovicTT** [comment]
An update:

TL;DR - I need to investigate why masks are calculated wrong for cross-chunk scenarios. See [ncvetkovic/11816_bh_reshuffle_rows](https://github.com/tenstorrent/tt-metal/tree/ncvetkovic/11816_bh_reshuffle_rows)

Detailed: I did a targeted test for embedding backwards op on BH. If I use a shape like [1, 32, 32, 32] which doesn't go across multiple chunks, I get a good result with some simple gradients. For instance, I have modified a test so that it visits only two rows in a larger embedding table - that works. Then I tried doing cross-chunk test (the latest commit on a branch). What seems to happen is that data for rows 32/33 ends up in rows 62/63 (so instead of mapping to rows 0/1 of chunk_1, it got mapped to rows 30/31 of chunck_1).

**Comment by dmakoviichuk-tt** [comment]
@ncvetkovicTT  thanks for update.  Does it mean shape like: [1,1, 1024, 2048] will not work? Or only shape where we have non divisible by 32 dimensions will not work?

**Comment by ncvetkovicTT** [comment]
Well not exactly, it means that all shapes larger than 32 won't work if they target rows from different tiles. Finding the root cause of this is my top priority so I hope to deliver a fix soon.

**Comment by ncvetkovicTT** [comment]
If I add some DPRINTs before `cb_wait_front(cb_out_intermed, max_tiles_per_core);` in the compute kernel (`embedding_backwards.cpp`), the test passes. However, adding TTI_NOP or asm("nop") still makes the test fail.

I am suspecting that there's a race condition somewhere when writing masks, since it seems like that in the test case I'm using, where I only have 2 chunks, the same mask is used for both, unless I add these DPRINT statements. Will dig deeper tomorrow.

**Comment by dmakoviichuk-tt** [comment]
I think we got same results @vmelnykovTT added some prints and it started working for some cases. But as I remember it freezed later anyway

**Comment by ncvetkovicTT** [comment]
Found a workaround for the race condition. Creating a PR tonight so that you can use the op. Maybe I will have to delay finding the root cause given that there are other P0 bugs waiting

**Comment by ncvetkovicTT** [comment]
@dmakoviichuk-tt Does any test besides `tt-metal/tests/ttnn/unit_tests/operations/test_backward_embedding.py` call `embedding_backwards.cpp` compute kernel? If not, we can skip post-commit and merge this ASAP:
https://github.com/tenstorrent/tt-metal/pull/28742

**Comment by dmakoviichuk-tt** [comment]
@ncvetkovicTT  we have a simple c++ test in tt-train too. But we didnt' enable blackhole tests yet.

**Comment by ncvetkovicTT** [comment]
Closed with https://github.com/tenstorrent/tt-metal/pull/28742
